packages/server/src/handlers/BaseHandler.ts

```diff
@@ -18,15 +23,20 @@ export class BaseHandler extends EventEmitter {
       throw new Error('Store must be defined')
     }
 
+    if (!options.locker) {
+      options.locker = new MemoryLocker()
+    }
```

Comment:
I don't see a reason why we wouldn't do this in the server constructor as that's the place where we check options. We can do something like this to type it correctly:

```ts
  options: WithRequired<ServerOptions, 'locker'>

  constructor(options: ServerOptions & {datastore: DataStore})
```

Reply:
there are a couple of reasons why i decided to not doing it on the Server constructor:

- 1) it means we need to add the `MemoryLocker` on all the tests around the code base that tests the BaseHandler or handlers

- 2) The TS definition seems way more ugly since we'll need the following:

```ts
//server.ts

export class Server extends EventEmitter {
  datastore: DataStore
  handlers: Handlers
  options: WithRequired<ServerOptions, 'locker'>

  constructor(options: ServerOptions & {datastore: DataStore}) {
    super()

    if (!options) {
      throw new Error("'options' must be defined")
    }

    if (!options.path) {
      throw new Error("'path' is not defined; must have a path")
    }

    if (!options.datastore) {
      throw new Error("'datastore' is not defined; must have a datastore")
    }

    if (!options.locker) {
      options.locker = new MemoryLocker()
    }

    const {datastore, ...rest} = options
 
    this.options = rest as WithRequired<ServerOptions, 'locker'>
  }
 ...
}
```

```ts
// baseHandler.ts
export class BaseHandler extends EventEmitter {
  options: WithRequired<ServerOptions, 'locker'>
  store: DataStore

  constructor(store: DataStore, options: WithRequired<ServerOptions, 'locker'>) {
    super()
    if (!store) {
      throw new Error('Store must be defined')
    }

    this.store = store
    this.options = options
  }
}
```

to be more ergonomic the handlers should have the default value, otherwise you cannot instantiate them without passing the locker in.

if this is what we want, then i can modify it as above

Reply:
Maybe I misunderstand but since you already made `locker` a required argument on the handlers, TS should already complain if you don't add them in tests? So moving that around wouldn't change anything there.

Indeed I don't think you can get rid of the `rest as WithRequired<ServerOptions, 'locker'>`, but what's more important I think is that that happens _once_, rather than every time you access `locker` somewhere in the program.

What do you think?

Reply:
> Maybe I misunderstand but since you already made locker a required argument on the handlers, TS should already complain if you don't add them in tests? So moving that around wouldn't change anything there.

No, because the constructor argument of the BaseHandler still uses `ServerOptions` which doesn't have the `locker` property required, the property `options` in the BaseHandler options needs to be `required`, this is why we can set the default value in the constructor if not provided.

But anyhow, I'll make the change and keep the default on the server constructor